record: TestSyncRecordWithSignalChan is flaky under race

https://github.com/cockroachdb/pebble/actions/runs/4169192744/jobs/7216870660

```
--- FAIL: TestSyncRecordWithSignalChan (0.00s)
    log_writer_test.go:221: 
        	Error Trace:	log_writer_test.go:221
        	Error:      	Not equal: 
        	            	expected: 4
        	            	actual  : 5
        	Test:       	TestSyncRecordWithSignalChan
```

--- a/record/log_writer_test.go
+++ b/record/log_writer_test.go
@@ -218,7 +218,11 @@ func TestSyncRecordWithSignalChan(t *testing.T) {
 		require.NoError(t, err)
 		syncWG.Wait()
 		require.NoError(t, syncErr)
-		require.Equal(t, cap(semChan)-(i+1), len(semChan))
+		// The waitgroup is released before the channel is read, so wait if
+		// necessary.
+		require.Eventually(t, func() bool {
+			return cap(semChan)-(i+1) == len(semChan)
+		}, 10*time.Second, time.Millisecond)
 	}
 }
 

